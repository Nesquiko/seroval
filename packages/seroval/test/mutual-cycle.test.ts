import { describe, it, expect } from 'vitest';
import {
  serialize,
  deserialize,
  ServerValue,
  AsyncServerValue,
  serializeAsync,
  fromJSON,
  toJSON,
  toJSONAsync,
} from '../src';

describe('mutual cyclic references', () => {
  describe('serialize', () => {
    it('supports Arrays and Arrays', () => {
      const a: ServerValue[] = [];
      const b: ServerValue[] = [];
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Arrays and Objects', () => {
      const a: ServerValue[] = [];
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Objects and Objects', () => {
      const a: Record<string, ServerValue> = {};
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
  });
  describe('serializeAsync', () => {
    it('supports Arrays and Arrays', async () => {
      const a: AsyncServerValue[] = [];
      const b: AsyncServerValue[] = [];
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
    it('supports Arrays and Objects', async () => {
      const a: AsyncServerValue[] = [];
      const b: Record<string, AsyncServerValue> = {};
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
    it('supports Objects and Objects', async () => {
      const a: Record<string, AsyncServerValue> = {};
      const b: Record<string, AsyncServerValue> = {};
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
  });
  describe('toJSON', () => {
    it('supports Arrays and Arrays', () => {
      const a: ServerValue[] = [];
      const b: ServerValue[] = [];
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Arrays and Objects', () => {
      const a: ServerValue[] = [];
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Objects and Objects', () => {
      const a: Record<string, ServerValue> = {};
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<ServerValue[][]>(result);
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Arrays and Arrays', async () => {
      const a: AsyncServerValue[] = [];
      const b: AsyncServerValue[] = [];
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
    it('supports Arrays and Objects', async () => {
      const a: AsyncServerValue[] = [];
      const b: Record<string, AsyncServerValue> = {};
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
    it('supports Objects and Objects', async () => {
      const a: Record<string, AsyncServerValue> = {};
      const b: Record<string, AsyncServerValue> = {};
      a[0] = Promise.resolve(b);
      b[0] = Promise.resolve(a);
      const example = [a, b];
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Promise<any>[][]>(result);
      expect(back[0]).toStrictEqual(await back[1][0]);
      expect(back[1]).toStrictEqual(await back[0][0]);
    });
  });
});
