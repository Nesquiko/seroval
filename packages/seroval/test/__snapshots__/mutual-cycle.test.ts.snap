// Vitest Snapshot v1

exports[`mutual cyclic references > serialize > supports Arrays and Arrays 1`] = `"((h,j,k)=>(k=[h=[j=[,]],j],j[0]=h,k))()"`;

exports[`mutual cyclic references > serialize > supports Arrays and Objects 1`] = `"((h,j,k)=>(k=[h=[j={}],j],j[0]=h,k))()"`;

exports[`mutual cyclic references > serialize > supports Objects and Objects 1`] = `"((h,j,k)=>(k=[h={0:j={}},j],j[0]=h,k))()"`;

exports[`mutual cyclic references > serializeAsync > supports Arrays and Arrays 1`] = `"((h,j)=>([h=[Promise.resolve(j=[Promise.resolve().then(()=>h)])],j]))()"`;

exports[`mutual cyclic references > serializeAsync > supports Arrays and Objects 1`] = `"((h,j)=>([h=[Promise.resolve(j={0:Promise.resolve().then(()=>h)})],j]))()"`;

exports[`mutual cyclic references > serializeAsync > supports Objects and Objects 1`] = `"((h,j)=>([h={0:Promise.resolve(j={0:Promise.resolve().then(()=>h)})},j]))()"`;

exports[`mutual cyclic references > toJSON > supports Arrays and Arrays 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":15,\\"i\\":1,\\"l\\":1,\\"a\\":[{\\"t\\":15,\\"i\\":2,\\"l\\":1,\\"a\\":[{\\"t\\":10,\\"i\\":1}]}]},{\\"t\\":10,\\"i\\":2}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,2]}"`;

exports[`mutual cyclic references > toJSON > supports Arrays and Objects 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":15,\\"i\\":1,\\"l\\":1,\\"a\\":[{\\"t\\":16,\\"i\\":2,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":10,\\"i\\":1}],\\"s\\":1}}]},{\\"t\\":10,\\"i\\":2}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,2]}"`;

exports[`mutual cyclic references > toJSON > supports Objects and Objects 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":16,\\"i\\":1,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":16,\\"i\\":2,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":10,\\"i\\":1}],\\"s\\":1}}],\\"s\\":1}},{\\"t\\":10,\\"i\\":2}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,2]}"`;

exports[`mutual cyclic references > toJSONAsync > supports Arrays and Arrays 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":15,\\"i\\":1,\\"l\\":1,\\"a\\":[{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":15,\\"i\\":3,\\"l\\":1,\\"a\\":[{\\"t\\":18,\\"i\\":4,\\"f\\":{\\"t\\":10,\\"i\\":1}}]}}]},{\\"t\\":10,\\"i\\":3}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,3]}"`;

exports[`mutual cyclic references > toJSONAsync > supports Arrays and Objects 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":15,\\"i\\":1,\\"l\\":1,\\"a\\":[{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":16,\\"i\\":3,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":18,\\"i\\":4,\\"f\\":{\\"t\\":10,\\"i\\":1}}],\\"s\\":1}}}]},{\\"t\\":10,\\"i\\":3}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,3]}"`;

exports[`mutual cyclic references > toJSONAsync > supports Objects and Objects 1`] = `"{\\"t\\":{\\"t\\":15,\\"i\\":0,\\"l\\":2,\\"a\\":[{\\"t\\":16,\\"i\\":1,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":16,\\"i\\":3,\\"d\\":{\\"k\\":[\\"0\\"],\\"v\\":[{\\"t\\":18,\\"i\\":4,\\"f\\":{\\"t\\":10,\\"i\\":1}}],\\"s\\":1}}}],\\"s\\":1}},{\\"t\\":10,\\"i\\":3}]},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[1,3]}"`;
